File: urbanairship-core/src/main/java/com/urbanairship/config/UrlBuilder.java
/* Copyright Airship and Contributors */

package com.urbanairship.config;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

/**
 * Url builder.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class UrlBuilder {

    private Uri.Builder uriBuilder;

    /**
     * Default constructor.
     *
     * @param url The URL.
     */
    public UrlBuilder(@Nullable String url) {
        if (url != null) {
            uriBuilder = Uri.parse(url).buildUpon();
        }
    }

    @NonNull
    public UrlBuilder appendEncodedPath(@NonNull String path) {
        if (uriBuilder != null) {
            uriBuilder.appendEncodedPath(path);
        }
        return this;
    }

    @NonNull
    public UrlBuilder appendPath(@NonNull String path) {
        if (uriBuilder != null) {
            uriBuilder.appendPath(path);
        }
        return this;
    }

    @NonNull
    public UrlBuilder appendQueryParameter(@NonNull String key, @NonNull String value) {
        if (uriBuilder != null) {
            uriBuilder.appendQueryParameter(key, value);
        }
        return this;
    }

    /**
     * Returns the Uri. If the Uri is malformed or the base Uri is null, null will be returned.
     *
     * @return The Uri or null.
     */
    @Nullable
    public Uri build() {
        if (uriBuilder == null) {
            return null;
        }
        return uriBuilder.build();
    }
}

File: urbanairship-core/src/main/java/com/urbanairship/config/UrlBuilder.kt
/* Copyright Airship and Contributors */
package com.urbanairship.config

import android.net.Uri
import androidx.annotation.RestrictTo

/**
 * Url builder.
 *
 * @hide
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class UrlBuilder public constructor(public val baseUrl: String?) {
    private var uriBuilder = baseUrl?.let { Uri.parse(it).buildUpon() }

    public fun appendEncodedPath(path: String): UrlBuilder {
        return this.also { uriBuilder?.appendEncodedPath(path) }
    }

    public fun appendPath(path: String): UrlBuilder {
        return this.also { uriBuilder?.appendPath(path) }
    }

    public fun appendQueryParameter(key: String, value: String): UrlBuilder {
        return this.also { uriBuilder?.appendQueryParameter(key, value) }
    }

    /**
     * Returns the Uri. If the Uri is malformed or the base Uri is null, null will be returned.
     *
     * @return The Uri or null.
     */
    public fun build(): Uri? = uriBuilder?.build()
}
